dss/stepfunctions/visitation/utils.py

```diff
@@ -0,0 +1,86 @@
+
+import time
+import typing
+import threading
+from .. import step_functions_list_executions, step_functions_describe_execution
+
+
+def current_walker_executions(sentinel_execution_name: str) -> typing.List[str]:
+    all_execs = step_functions_list_executions(
+        'dss-visitation-walker-{stage}',
+        status_filter='RUNNING'
+    )
+
+    execs = [e for e in all_execs
+             if e['name'].endswith(sentinel_execution_name)]
+
+    return execs
+
+
+def throttled_iter(iterable: typing.Iterable, calls_per_second: int=2, bunch_size: int=10) -> typing.Iterable:
+    t = time.time()
+    k = 0
+
+    for item in iterable:
+
+        if k >= bunch_size:
+            dt = time.time() - t
+            if k / dt > calls_per_second:
+                wait_time = k / calls_per_second - dt
+                time.sleep(wait_time)
+
+            t = time.time()
+            k = 0
+
+        k += 1
+        yield item
+
+
+def walker_execution_name(sentinel_name: str, prefix: str) -> str:
+    return '{}----{}'.format(
+        prefix.replace('/', '____'),
+        sentinel_name
+    )
+
+
+def walker_prefix(walker_execution_name: str) -> str:
+    encoded_prefix, sentinel_execution_name = walker_execution_name.split('----')
+    prefix = encoded_prefix.replace('____', '/')
+    return prefix
+
+
+class Timeout:
```

Comment:
Consider using signals instead. They avoid shutdown issues, and you will still be called if a thread is hogging the GIL.

```
def alarm_handler(signum, frame):
        raise MyTimeout("...")

    signal.signal(signal.SIGALRM, alarm_handler)
    timeout_seconds = (context.get_remaining_time_in_millis() / 1000) - 10
    context.log("Setting timeout to {}".format(timeout_seconds))
    signal.alarm(timeout_seconds)

    try:
        ... do work ...
    except MyTimeout:
        ... shutdown ...
```

Reply:
Thanks, I'll use this, or check for execution time within the loop.